ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,215 @@
+      totalValidSize += value;
+    }
+    return new Pair<>(logSegmentValidSizeResult.getFirst(), totalValidSize);
+  }
+
+  @Override
+  public Pair<Long, Map<String, Long>> getValidSizeByLogSegment(TimeRange timeRange) throws StoreException {
+    long deleteReferenceTimeInSecs = timeRange.getEndTimeInSecs();
+    long expirationReferenceTimeInMs = time.milliseconds();
+    Map<String, Long> validSizePerLogSegment =
+        collectValidSizeByLogSegment(deleteReferenceTimeInSecs * Time.MsPerSec, expirationReferenceTimeInMs);
+    return new Pair<>(deleteReferenceTimeInSecs, validSizePerLogSegment);
+  }
+
+  @Override
+  public Map<String, Map<String, Long>> getValidDataSizeByContainers() throws StoreException {
+    long referenceTimeInMs = time.milliseconds();
+    return collectValidDataSizeByContainer(referenceTimeInMs);
+  }
+
+  /**
+   * Walk through the entire index and collect valid data size information per container.
+   * @param referenceTimeInMs the reference time to be used to decide whether or not a blob is valid
+   * @return a nested {@link Map} of serviceId to containerId to valid data size
+   */
+  private Map<String, Map<String, Long>> collectValidDataSizeByContainer(long referenceTimeInMs) throws StoreException {
+    Set<String> deleteKeys = new HashSet<>();
+    Map<String, Map<String, Long>> validDataSizePerContainer = new HashMap<>();
+    for (IndexSegment indexSegment : index.getIndexSegments().descendingMap().values()) {
+      diskIOScheduler.getSlice(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, BlobStoreStats.IO_SCHEDULER_JOB_ID, 1);
+      List<MessageInfo> messageInfos = new ArrayList<>();
+      try {
+        indexSegment.getEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE), messageInfos,
+            new AtomicLong(0));
+      } catch (IOException e) {
+        throw new StoreException("I/O error while collecting valid data size per container", e, StoreErrorCodes.IOError);
+      }
+      populateContainerMap(validDataSizePerContainer, referenceTimeInMs, messageInfos, deleteKeys);
+    }
+
+    return validDataSizePerContainer;
+  }
+
+  /**
+   * Walk through the entire index and collect valid size information per log segment.
+   * @param deleteReferenceTimeInMs the reference time to be used to decide whether or not a blob is deleted
+   * @param expirationReferenceTimeInMs the reference time to be used to decide whether or not a blob is expired
+   * @return a {@link Map} of log segment name to valid size
+   */
+  private Map<String, Long> collectValidSizeByLogSegment(long deleteReferenceTimeInMs, long expirationReferenceTimeInMs)
+      throws StoreException {
+    Set<String> deleteKeys = new HashSet<>();
+    Map<String, Long> validSizePerLogSegment = new HashMap<>();
+    for (IndexSegment indexSegment : index.getIndexSegments().descendingMap().values()) {
+      diskIOScheduler.getSlice(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, BlobStoreStats.IO_SCHEDULER_JOB_ID, 1);
```

Comment:
it looks like lines 119-126 could be a function since the same code is repeated in lines 94-101. Are there any differences?